connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SubmittedRecords.java

```diff
@@ -68,6 +75,9 @@ SubmittedRecord submit(Map<String, Object> partition, Map<String, Object> offset
         SubmittedRecord result = new SubmittedRecord(partition, offset);
         records.computeIfAbsent(result.partition(), p -> new LinkedList<>())
                 .add(result);
+        synchronized (this) {
+            numUnackedMessages.incrementAndGet();
+        }
```

Comment:
It definitely feels like an anti-pattern to synchronize around modifications to an atomic field. I've gone into further detail below but TL;DR: this is a combination of wanting to avoid a race condition (hence the synchronized block) and wanting to silence SpotBugs warnings (hence the use of an `AtomicInteger` instead of a `volatile int`).